Initial manifest creation steps

matthewrsj opened the pull request:
this is still a WIP, but I'm opening a PR so people can comment on the methods.

matthewrsj committed:
Add copyright notice to swupdlib files

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
Add initial manifest creation steps

Adds tests to most added functions as well.

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

cmarcelo commented on Makefile.gobuild:
```diff
@@ -1,7 +1,7 @@
 CUR_DIR = $(shell pwd)
 
```
To account for the case GOPATH is empty (which go tools default to $HOME/go), add something like

    GOPATH ?= ${HOME}/go

here in the top of the Makefile.

cmarcelo commented on Makefile.gobuild:
```diff
@@ -1,7 +1,7 @@
 
 %.build:
-	GOPATH=$(CUR_DIR) go install $(subst .build,,$@)
+	GOPATH=$(GOPATH):$(CUR_DIR) go install $(subst .build,,$@)
```
Change the order, the first entry in GOPATH is used to figure out where to put the binaries, in this Makefile we want it to be in the "cur dir" / bin.

matthewrsj replied:
This is actually pre-empted by an upstream change.

cmarcelo commented on src/swupd/config.go:
```diff
@@ -0,0 +1,144 @@
+	"strconv"
+	"strings"
+
+	"github.com/go-ini/ini"
```
We should use dep to vendor this. And since the upstream does proper releases, stick to a released tag.

(If you do that, you don't even need to make the changes to Makefile anymore)

matthewrsj replied:
Yep, haven't updated this branch to reflect the decision to use dep yet.

cmarcelo commented on src/swupd/config.go:
```diff
@@ -0,0 +1,144 @@
+var outputDir = "/var/lib/update/www/"
+var debuginfoBanned = true
+var debuginfoLib = "/usr/lib/debug"
+var debuginfoSrc = "/usr/src/debug"
```
Suggestion: use a struct to pass around config information.

Create a type which is a struct with all these attributes.
Create a var from that type with those default values.


    type config struct {
        emptyDir string
        // ...
    }
    var DefaultConfig = config{
        emptyDir: "/var/lib/update/empty",
        // ...
    }

Change the read function to return a Config, which you can initialize with DefaultConfig and fill in as you read the INI file.

This will make testing more straightforward. Avoiding the global means passing Config around more, if that is unfeasible you can still keep a global pointer to the current configuration.

matthewrsj replied:
good idea

cmarcelo commented on src/swupd/config.go:
```diff
@@ -0,0 +1,144 @@
+		return err
+	}
+
+	if ed := cfg.Section("Server").Key("emptydir").Value(); ed != "" {
```
This "steals" the empty string to represent key doesn't exist in the configuration file. We may be bitten in the future by some value that could be set to empty (to avoid some non-empty default). Suggest you use something like:

    if key, err := cfg.Section("Server").GetKey("emptydir"); err == nil {
        result.emptyDir = key.Value()
    }

cmarcelo commented on src/swupd/config_test.go:
```diff
@@ -0,0 +1,107 @@
+		debuginfoBanned = defaultDebuginfoBanned
+		debuginfoLib = defaultDebuginfoLib
+		debuginfoSrc = defaultDebuginfoSrc
+	}()
```
Avoiding the use of the global in the functions you avoid this kind of state recovery in the tests.

cmarcelo commented on src/swupd/create_manifests.go:
```diff
@@ -0,0 +1,238 @@
+)
+
+var StateDir = "/var/lib/swupd"
+var MinVersion = false
```
Optional: like the suggestion for the configuration, try to stick the global variables into structs when possible (and when the globals are related). It at least make easier to later avoid using globals in the functions by passing the struct around.

cmarcelo commented on src/swupd/create_manifests.go:
```diff
@@ -0,0 +1,238 @@
+	timeStamp   time.Time
+}
+
+func initBuildEnv() error {
```
Optional: "resetBuildEnv"?

cmarcelo commented on src/swupd/create_manifests.go:
```diff
@@ -0,0 +1,238 @@
+//            ...
+//    LAST_VER (file containing "0\n" for now)
+func initBuildDirs(version uint32, groups []string) error {
+	// what is this???
```
What?

matthewrsj replied:
sorry, placeholder to remind me to do more research as to why LAST_VER file is always initialized to `0`. This PR is still WIP

cmarcelo commented on src/swupd/create_manifests.go:
```diff
@@ -0,0 +1,238 @@
+			return newManifests, err
+		}
+
+		newM := Manifest{
```
Optional: use &Manifest{...} to already create a pointer type.

cmarcelo commented on src/swupd/create_manifests.go:
```diff
@@ -0,0 +1,238 @@
+var StateDir = "/var/lib/swupd"
+var MinVersion = false
+
+type Update struct {
```
Optional: "UpdateInfo"?

cmarcelo commented on src/swupd/create_manifests.go:
```diff
@@ -0,0 +1,238 @@
+			return newManifests, err
+		}
+
+		// apply heuristics
```
Optional: either remove the comment (since the text is exactly like the function call) or elaborate a bit ("heuristics to achieve what?").

matthewrsj replied:
pseudo code still hanging around

cmarcelo commented on src/swupd/create_manifests.go:
```diff
@@ -0,0 +1,238 @@
+
+		// read in bundle includes
+		if err := newM.readIncludes(); err != nil {
+			return newManifests, err
```
Should we just return an empty list when there was an error, or you will take advantage of the partial result?

matthewrsj replied:
My thought was in the future we may use the partial result or log it. Right now I do nothing with it.

cmarcelo commented on src/swupd/config.go:
```diff
@@ -0,0 +1,144 @@
+)
+
+var emptyDir = "/var/lib/update/empty/"
+var imageBase = "/var/lib/update/image/"
```
Optional: maybe that is not the right time, but I think this should be called chrootBaseDir or something like. The word "image" has a different meaning that sometimes cause confusion. I'm fine keeping the actual directory as is ("image/") for now to avoid any incompatibility.

cmarcelo commented on src/swupd/create_manifests.go:
```diff
@@ -0,0 +1,238 @@
+	}
+
+	oldFullManifest := Manifest{}
+	oldFullManifestPath := filepath.Join(StateDir, string(lastVersion), "Manifest.full")
```
Is string(lastVersion) doing the right thing? A quick test here with small values of uint32 gives me strange results.

Consider using strconv functions, and if the conversion repeats a lot, do it once and reuse it.

matthewrsj replied:
yeah, it absolutely isn't. I've since updated it to use `fmt.Sprint()` instead.

cmarcelo commented on src/swupd/filesystem.go:
```diff
@@ -0,0 +1,91 @@
+		fh = Hashcalc(file.Name)
+	case mode.IsDir():
+		file.Type = typeDirectory
+		fh = "19cb945a2ce72a00335960767c100f27cbdf0e7e58664ae24e856870be117ccd"
```
How was this constant calculated?

Optional: make the calculation explicit so if the Hash algorithm changes, this will change too.

matthewrsj replied:
ok, this was a hard-coded test value. I was still waiting for the hash calculation function to be complete. I hope to be pushing an update today with much more complete code (working).

cmarcelo commented on src/swupd/filesystem.go:
```diff
@@ -0,0 +1,91 @@
+		return false
+	}
+
+	return true
```
Maybe more readable:

    return info.Mode().IsDir() && info.Mode().Perm() == perm

matthewrsj replied:
Cleaner, yes. I don't know about more readable.

cmarcelo commented on src/swupd/manifest.go:
```diff
@@ -39,9 +42,11 @@ type ManifestHeader struct {
+	Header       ManifestHeader
+	Files        []*File
+	DeletedFiles []*File
+	Changed      bool
```
What does Changed stands for?

matthewrsj replied:
Looks like I don't use that like I thought I would. It was intended to be a boolean representing whether the manifest needed to be regenerated or not. I don't need that any more because I append changed files to the `newManifests` slice.

cmarcelo commented on src/swupd/manifest.go:
```diff
@@ -333,3 +343,154 @@ func (m *Manifest) WriteManifestFile(path string) error {
+	return changed
+}
+
+func (m *Manifest) filesAdded(oldManifest *Manifest) int {
```
Optional: countFilesAdded? Same suggestion for the function below.

cmarcelo reviewed (commented):
Couldn't finish the entire PR today, will continue tomorrow. But there's enough useful feedback.

matthewrsj committed:
Update swupd library to be gometalinter compliant

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
Fix up several tests for Travis compatibility and improved debugging

This includes manually setting test file permissions to be consistent
with what we expect and renaming test directories to more test-specific
names.

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj commented:
This is ready for review now @cmarcelo @icarus-sparry @tmarcu @phmccarty 

Travis is failing because of a hash mismatch for manifest read/write test - I believe this is because the file permissions differ when running on travis (I set the file permissions manually to what I saw on my system). What permissions should a manifest that we write have?

Travis also fails because of gometalinter errors on the rest of the repo, this part can be ignored.

phmccarty commented:
@matthewrsj In the swupd-client and swupd-server travis configs, we set the umask manually when running `make check`, because Ubuntu's default umask is different than on Clear Linux. Is that the issue you're running into?

cmarcelo commented on .travis.yml:
```diff
@@ -11,6 +11,7 @@ before_install:
     - go get -u gopkg.in/alecthomas/gometalinter.v1
     - gometalinter.v1 --install
     - go get -u github.com/go-ini/ini
+    - go get -u github.com/matthewrsj/mixer-tools/swupd
```
What error were you getting to need this?

It seems wrong to clone the master repo again when testing a specific PR. If I understand it correctly the sources we need are already in $GOPATH/github.com/clearlinux/mixer-tools/.

matthewrsj replied:
`swupd` is not currently in clearlinux/mixer-tools so the fullfiles test program failed to compile.

cmarcelo replied:
The way we are running Travis, it is getting the content of the PR (which contains swupd directory) and puts it inside a directory $GOPATH/src/github.com/clearlinux/mixer-tools, as specified by go_import_path https://docs.travis-ci.com/user/languages/go/#Go-Import-Path.

cmarcelo replied:
About the tests being executed: the only change we are missing is adding go test call to travis script (as you did in the other patch), but the tests should run just fine: e.g. the PR #15 added the call to go test in travis, and resulted in

    https://travis-ci.org/matthewrsj/mixer-tools/builds/321963818?utm_source=github_status&utm_medium=notification

note how the tests are correctly executed.

matthewrsj replied:
Yep, you are right. Didn't catch the go_import_path thing. Thanks for explaining.

cmarcelo commented on swupd/config.go:
```diff
@@ -18,56 +18,88 @@ import (
+}
+
+func getConfig() config {
+	setDefaultConfig()
```
I've noticed that you changed the defaults to be parametrized based on StateDir -- this wasn't present in the previous code. Can people change StateDir, call getConfig and _expect_ the default values to change?

If no: remove this call and initialize defaultConfig using a struct literal

    var defaultConfig = config{
        imageBase: filepath.Join(...)
    }

If yes: you are essentially passing parameters via a global variable, should avoid that and pass parameters explicitly to the functions that need them. This is fragile. We can fix this later, though.

matthewrsj replied:
You are right, this is a little messy. I will fix it up.

cmarcelo commented on swupd/config.go:
```diff
@@ -18,56 +18,88 @@ import (
-func readServerINI(path string) error {
+// StateDir is the directory under which swupd will create the update
+// defaults to /var/lib/swupd unless overridden
+var StateDir = "/var/lib/swupd"
```
Why not part of the Config?

cmarcelo commented on swupd/manifest.go:
```diff
@@ -311,21 +311,15 @@ func writeManifestFileEntry(file *File, w *bufio.Writer) error {
 	var err error
 	var f *os.File
-	if f, err = os.Create(path); err != nil {
+	if f, err = os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644); err != nil {
```
Minor: don't you also need os.O_TRUNC?

cmarcelo commented on swupd/manifest_test.go:
```diff
@@ -275,19 +275,15 @@ func TestWriteManifestFile(t *testing.T) {
-	defer os.Remove(f.Name())
-
-	newpath := f.Name()
+	// do not use a tempfile here, we just need the unique name
```
Minor: this is a change in the wrong direction IMHO. If this test is copy/pasted to create a new one (a reasonable way to create a new test), we will have a problem if they run in parallel.

matthewrsj replied:
Maybe, but all we need in that case is a unique name. The function takes a pathname, not a file, so creating a file here is not necessary. Is it really important to protect against future copy-paste issues now?

matthewrsj commented:
Do not review yet, tests and changes incoming.

matthewrsj committed:
Add several new tests for manifest creation

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
Fix several issues identified by tests

* missing debuginfo checks
* track manifest file count
* perform format bump checks
* track and deprecate ghosted files

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
Improve configuration handling

Instead of leaving StateDir hanging around as a global, include it as
part of the config struct that is passed around.

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

cmarcelo commented on swupd/config.go:
```diff
@@ -0,0 +1,178 @@
+
+	cfg, err := ini.InsensitiveLoad(path)
+	if err != nil {
+		// server.ini exists, but we were unable to read it
```
We should return error here, Load might be failing to due a parser error, so the user might think that the file is OK, but the program proceeds with a completely different configuration.

This functional should return (config, error).

cmarcelo commented on swupd/config_test.go:
```diff
@@ -0,0 +1,92 @@
+)
+
+func TestReadServerINI(t *testing.T) {
+	if c := readServerINI("nowhere", "noINI"); c != defaultConfig {
```
To compare structs we need to use reflect.DeepEqual. We want to test if the values are the same.

cmarcelo commented on swupd/config_test.go:
```diff
@@ -0,0 +1,92 @@
+	var err error
+	if _, err = readGroupsINI("nowhere"); err == nil {
+		// readGroupsINI does raise an error when the groups.ini file does not exist
+		// because it is required for the build
```
Optional: this comment could be in the readGroupsINI itself.

cmarcelo commented on swupd/create_manifests.go:
```diff
@@ -0,0 +1,260 @@
+import (
+	"errors"
+	"fmt"
+	//"io/ioutil"
```
Optional: remove old/commented import.

cmarcelo commented on swupd/create_manifests.go:
```diff
@@ -0,0 +1,260 @@
+)
+
+// MinVersion indicates a minimum version build in which all content must be generated
+var MinVersion = false
```
Since this PR don't use it yet, don't add it. When we make use of it, we can avoid this information being a global.

cmarcelo commented on swupd/create_manifests.go:
```diff
@@ -0,0 +1,260 @@
+		oldM := &Manifest{}
+		oldMPath := filepath.Join(c.outputDir, fmt.Sprint(ui.lastVersion), "Manifest."+bundle)
+		if err := oldM.ReadManifestFromFile(oldMPath); err != nil {
+			// old manifest may not exist, continue
```
If we are ignoring the error, just ignore the result of ReadManifestFromFile.

Question: if the manifest exists but is malformed, should we still ignore it or should we fail processing?

cmarcelo commented on swupd/create_manifests.go:
```diff
@@ -0,0 +1,260 @@
+		deleted := newM.newDeleted(oldM)
+		if changedFiles == 0 && added == 0 && deleted == 0 && !changedIncludes {
+			newM.Header.Version = oldM.Header.Version
+			continue
```
If we are continuing here, do we need to set newM version? It seems it will not be used anymore since it is not in the list of newManifests.

cmarcelo commented on swupd/create_manifests.go:
```diff
@@ -0,0 +1,260 @@
+	oldFullManifestPath := filepath.Join(c.outputDir, fmt.Sprint(lastVersion), "Manifest.full")
+	if err = oldFullManifest.ReadManifestFromFile(oldFullManifestPath); err != nil {
+		// might not exist, so the empty manifest is fine
+	}
```
If ignoring the result, you can make the call without assigning to err. Keep the comment around.

(My observation about malformed vs. non-existent remains).

cmarcelo commented on swupd/create_manifests_test.go:
```diff
@@ -0,0 +1,376 @@
+	"testing"
+)
+
+func dirExistsWithPerm(path string, perm os.FileMode) bool {
```
Optional: for some helpers used in tests to verify a condition that must hold, it is a good idea to pass "t *testing.T" and let the helper call t.Fatal directly. If you are always failing a test after calling the helper with a message from data passed to the helper, then it is a good candidate for passing t and letting it fail for you.

It is common to see the prefix "must" being used to make clear this is a condition that must hold. "mustDirExistsWithPerm".

cmarcelo commented on swupd/create_manifests_test.go:
```diff
@@ -0,0 +1,376 @@
+		return false
+	}
+
+	return strings.Contains(string(b), sub)
```
Optional: there is a bytes.Contains, so you would convert the smaller input instead of the potentially larger one.

cmarcelo commented on swupd/create_manifests_test.go:
```diff
@@ -0,0 +1,376 @@
+		return ""
+	}
+
+	return re.FindString(string(b))
```
Optional: "return string(re.Find(b))"

cmarcelo commented on swupd/create_manifests_test.go:
```diff
@@ -0,0 +1,376 @@
+		t.Fatalf("Unable to remove %s", testDir)
+	}
+
+	cmd := exec.Command("cp", "-a", testDir+".bak", testDir)
```
Create a copyDir(t, dirA, dirB) helper function to take the noise a bit out of the body of the tests.

Optional: the prefix .bak sounds like "backup" and gives the wrong message, as these directories contain the input of the tests.

cmarcelo commented on swupd/create_manifests_test.go:
```diff
@@ -0,0 +1,376 @@
+	}
+
+	if err := CreateManifests(10, false, 1, testDir); err != nil {
+		t.Error(err)
```
Optional: should this simply be t.Fatal?

cmarcelo commented on swupd/create_manifests_test.go:
```diff
@@ -0,0 +1,376 @@
+	}
+
+	if err := CreateManifests(10, false, 1, testDir); err != nil {
+		t.Error(err)
```
Optional: I would go t.Fatal, since the next one is likely to fail if this one fails.

I don't think having multiple errors in a test result help much, since you'll have to start with the first (given the other errors might be a consequence of the first). Note that the comment does not apply to table tests, which are effectively a set of tests.

cmarcelo commented on swupd/create_manifests_test.go:
```diff
@@ -0,0 +1,376 @@
+	testDir := "./testdata/testdata-filename-debuginfo"
+	if err := os.RemoveAll(testDir); err != nil {
+		t.Fatal("unable to remove " + testDir)
+	}
```
Optional: create a helper mustRemoveAll(t, dir).

cmarcelo commented on swupd/create_manifests_test.go:
```diff
@@ -0,0 +1,376 @@
+	}
+
+	if err := CreateManifests(20, true, 1, testDir); err == nil {
+		t.Error("CreateManifests successful when decrementing format")
```
Optional: explicitly mention that failure was expected.

cmarcelo commented on swupd/files.go:
```diff
@@ -221,3 +217,43 @@ func (f *File) getFlagString() (string, error) {
+	// file -> directory OK
+	// link -> file OK
+	// link -> directory OK
+	// directory -> anything TYPE CHANGE
```
Optional: since some type changes are OK, the name of this function could be "isUnsupportedTypeChanged".

cmarcelo commented on swupd/filesystem.go:
```diff
@@ -0,0 +1,109 @@
+)
+
+func filenameBlacklisted(fname string) bool {
+	illegalChars := ";&|*`/<>\\\"'"
```
Optional: use const instead of a variable.

cmarcelo commented on swupd/filesystem.go:
```diff
@@ -0,0 +1,109 @@
+
+	if filenameBlacklisted(filepath.Base(fname)) {
+		fmt.Fprintf(os.Stderr, "%s is a blacklisted file name\n", fname)
+		return nil
```
Optional: this should be an specific error (that can be identified by the caller) or a separate output value. Ideally we should let the caller decide where to print this.

cmarcelo commented on swupd/filesystem.go:
```diff
@@ -0,0 +1,109 @@
+		file.Type = typeLink
+	default:
+		fmt.Fprintf(os.Stderr, "%v is an unsupported file type\n", file.Name)
+		return nil
```
Optional: should be an error.

cmarcelo commented on swupd/filesystem.go:
```diff
@@ -0,0 +1,109 @@
+		return err
+	}
+
+	return nil
```
Replace last 5 lines with 'return err'.

cmarcelo commented on swupd/filesystem_test.go:
```diff
@@ -0,0 +1,73 @@
+
+func TestCreateFileFromPath(t *testing.T) {
+	Hashes = []*string{}
+	invHash = make(map[string]Hashval)
```
Call resetHash function.

cmarcelo commented on swupd/filesystem_test.go:
```diff
@@ -0,0 +1,73 @@
+		t.Error(err)
+	}
+
+	if len(m.Files) == 0 {
```
Check for the expected number of files. If we expect 2 and get 1, it would pass the zero test.

cmarcelo commented on swupd/hash_test.go:
```diff
@@ -138,7 +138,7 @@ func TestHashcalc(t *testing.T) {
 		result   string
 	}{
-		{"/", directoryhash},
+		//{"/", directoryhash},
```
Optional: Should this be removed? If not add a TODO.

matthewrsj replied:
Forgot about this. @icarus-sparry I'm getting a different hash for directories than the tests specify. I'm getting 6bbfee41be037199188871b4fb123de5c95b6dac6a597cabcd86a751158c1111 when the tests are expecting 6c27df6efcd6fc401ff1bc67c970b83eef115f6473db4fb9d57e5de317eba96e

matthewrsj replied:
disregard

cmarcelo commented on swupd/manifest.go:
```diff
@@ -319,3 +342,215 @@ func (m *Manifest) WriteManifestFile(path string) error {
+	return added
+}
+
+func (m *Manifest) newDeleted(oldManifest *Manifest) int {
```
Optional: filesAdded, newDeleted and linkPeersAndChanged could be calculated in the same pass using two sorted list of files, like the original swupd-server did (and this code does elsewhere).

cmarcelo commented on swupd/manifest.go:
```diff
@@ -319,3 +342,215 @@ func (m *Manifest) WriteManifestFile(path string) error {
+	return false
+}
+
+func (m *Manifest) subtractManifestFromManifest(m2 *Manifest) {
```
Optional: since it contains some non-trivial logic, this function is a good candidate for having tests.

cmarcelo reviewed (approved):
LGTM with a few comments, and a bunch of other optional comments.

I'm not a fan of creating directories our repo directory with temporary content for the tests unless really necessary, I rather using temporary directories (or one big temporary directory with named subdirectories inside).

Another thing I'm trying to find a better solution is for tests that are "splitted": the input is in one place and the expected output is in another -- this tend to make things harder. I'll give more thought about this and follow up.

matthewrsj committed:
Incorporate suggested changes

Make suggested changes from review of PR #9 in matthewrsj/mixer-tools

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj commented:
There is a single test that is failing in Travis that passes for me locally:
```
--- FAIL: TestCreateManifests (0.28s)
	create_manifests_test.go:142: 'filecount:	5' not found in 10/Manifest.test-bundle
	create_manifests_test.go:142: '10	/usr/share' not found in 10/Manifest.test-bundle
	create_manifests_test.go:172: 'filecount:	5' not found in 20/Manifest.test-bundle
```
I don't think this should hold up the merge of this PR, so I opened an issue for it (#20).

@cmarcelo I addressed all your non-optional comments and almost all your optional comments. TYVM for your detailed review. I'm merging this now.
